README.md
classic.models
==============

This package introduces includes datasets from the Classic Models database for the Analytics Designer from OpenText. The goal of this package is to provide a dataset for learning to integrate R scripts with the BIRT report design. These design files are used by the OpenText Information Hub (iHub) server. 

This package is a work in progress which includes:

* examples of grabbing tables of data from the iHub 
* examples of sending data to the iHub
* datasets from the Classic Models data

Future plans include adding:

* more R examples
* move examples from demos to vignettes
* more error checking

### Installation 
Install the package from the zip or the tar.gz file. For example:

```
install.packages("C:/Users/yourname/Downloads/classic.models_0.1.0.zip",repos = NULL) 
```

then, to use the package in your R workspace use the following command:

```
library(classic.models)
```

When you are ready to remove it, use the following command:

```
remove.packages("classic.models")
```

You can also install it directly from GitHub.
```
install.packages("devtools")
library("devtools")
install_github("shawngiese/classic.models")
```

If you receive any errors about loading the curl package, install that package manually. Then restart the installation of classic.models.
```
install.packages("curl")
```


### Notes
Building the package used some of the following resources:

* http://thecoatlessprofessor.com/programming/creating-an-r-data-package/
* https://cran.r-project.org/doc/manuals/R-exts.html#Data-in-packages
* https://github.com/search?p=5&q=user%3Acran+.onLoad+data&type=Code&utf8=✓
* http://tinyheero.github.io/jekyll/update/2015/07/26/making-your-first-R-package.html
* http://r-pkgs.had.co.nz/data.html
* https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html
* http://kbroman.org/pkg_primer/pages/docs.html

Don't forget devtools::build_vignettes()


NEWS
Don't forget devtools::build_vignettes() 

Converting date strings to date objects
data("ORDERS")
ORDERS$ORDERDATE <- as.Date(ORDERS$ORDERDATE, format= "%Y-%m-%d")
ORDERS$REQUIREDDATE <- as.Date(ORDERS$REQUIREDDATE, format= "%Y-%m-%d")
ORDERS$SHIPPEDDATE <- as.Date(ORDERS$SHIPPEDDATE, format= "%Y-%m-%d")
save(ORDERS, file="data/ORDERS.rda")

Other stuff todo:
* show version along with data file name
* change ; to : in first connection info question
* add some simple features to check connectivity to server
* add file check for configuration settings
* add markdown vignettes


demo/plotdata.r
# Create a bar plot
# This demo merges two data sets and aggregates the data.

mydata <- merge(ORDERDETAILS, ORDERS, "ORDERNUMBER")
shipping = aggregate(PRICEEACH*QUANTITYORDERED ~ STATUS, mydata, sum)
colnames(shipping) <- c("Status", "Value")
readline("press any key to continue")

#Display table
print(shipping)

# Set scipen to disable scientific notation in axis.
# Not needed in this example because we create our own axis.
# options(scipen=999)

# Create the bar plot and attach the axis with labels.
myplot = barplot(shipping$Value, names.arg=shipping$Status, ylim = c( 0 , 10000000 ), xlab="Status", yaxt="n")
my.axis <- paste("$",prettyNum(axTicks(2)/1000, big.mark = ","),"K",sep=" ")
axis(2,at=axTicks(2), labels=my.axis, las=1)
my.bars <-paste("$",prettyNum(round(shipping$Value/1000), big.mark = ","),"K",sep=" ")
text(myplot, shipping$Value, labels=my.bars, pos=3, offset=.1)



demo/upload_file.r
# R script to load file to iHub
# SG 5/2017

#Load libraries
require(ggplot2)
require(curl)
require(jsonlite)
require(httr)
load("ihub_connect.rdata")

readline("press any key to continue")

#Make a PDF file
data("mtcars")
pdf("test.pdf")
myplot <- ggplot(mtcars, aes(wt, mpg)) + geom_point()
print(myplot)
dev.off()
readline("press any key to continue")

#Set login information
login <- list(
  username = connect$ihub.user,
  password = connect$ihub.password
)
readline("press any key to continue")

#Get authToken - add verbose() as last encode for debugging
url.login <- paste("http://",connect$ihub.server,":8000/api/v2/login",sep="")
req <- httr::POST(url.login,
  body = login,
  encode = "form"
);
token <- paste(content(req)$authToken)
readline("press any key to continue")

#Upload file with authToken to volume root directory
#mime type is guessed from extension, otherwise add the mimetype
# upload_file("test.pdf", "application/pdf"))
url.upload <- paste("http://",connect$ihub.server,":8000/api/v2/files/upload",sep="")
req2 <- httr::POST(url.upload,
  httr::add_headers(
    "AuthToken" = token,
    "Content-Type" = "multipart/form-data",
	"Accept-Encoding" = "gzip, deflate",
    "Accept" = "application/json"
  ),
  body=list(name="test3.pdf", file=upload_file("test.pdf"))
);
print(req2)

demo/download_dataset.r
# R script to load data from iHub
# SG 5/2017

#Load libraries
require(curl)
require(jsonlite)
require(httr)
load("ihub_connect.rdata")

#Set login information
login <- list(
  username = connect$ihub.user,
  password = connect$ihub.password
)
readline("press any key to continue")

#Get authToken - add verbose() as last encode for debugging
url.login <- paste("http://",connect$ihub.server,":8000/api/v2/login",sep="")
req <- httr::POST(url.login,	
  body = login,
  encode = "form"
);
token <- paste(content(req)$authToken)


#Get data with authToken
url.dataobjects <- paste("http://",connect$ihub.server,":8000/api/v2/dataobjects/15000000100/Customers",sep="")
req2 <- httr::GET(url.dataobjects, add_headers(AuthToken = token))
json <- httr::content(req2, as = "text")
mydata <- fromJSON(json)

customer_data_frame = mydata[['data']]
readline("press any key to continue")

head(customer_data_frame)
readline("press any key to continue")

str(customer_data_frame)

vignettes/introduction.R
## ---- eval=FALSE---------------------------------------------------------
#  ORDERS$ORDERDATE <- as.Date(ORDERS$ORDERDATE, format= "%Y-%m-%d")
#  ORDERS$REQUIREDDATE <- as.Date(ORDERS$REQUIREDDATE, format= "%Y-%m-%d")
#  ORDERS$SHIPPEDDATE <- as.Date(ORDERS$SHIPPEDDATE, format= "%Y-%m-%d")
#  PAYMENTS$PAYMENTDATE <- as.Date(PAYMENTS$PAYMENTDATE, format= "%Y-%m-%d")

## ---- eval=FALSE---------------------------------------------------------
#  data <- merge(ORDERDETAILS, ORDERS, "ORDERNUMBER")

## ---- eval=FALSE---------------------------------------------------------
#  plot(data$ORDERDATE, data$PRICEEACH)

## ---- eval=FALSE---------------------------------------------------------
#  shipping = aggregate(PRICEEACH*QUANTITYORDERED ~ STATUS, data, sum)
#  colnames(shipping) <- c("Status", "Value")
#  # options(scipen=999)
#  myplot = barplot(shipping$Value, names.arg=shipping$Status, ylim = c( 0 , 10000000 ), xlab="Status", yaxt="n")
#  my.axis <- paste("$",prettyNum(axTicks(2)/1000, big.mark = ","),"K",sep=" ")
#  axis(2,at=axTicks(2), labels=my.axis, las=1)
#  my.bars <-paste("$",prettyNum(round(shipping$Value/1000), big.mark = ","),"K",sep=" ")
#  text(myplot, shipping$Value, labels=my.bars, pos=3, offset=.1)



R/ihub_functions.r
#' Collecting information to connect to your server
#'
#' The default Information Hub connection is to http://localhost
#' with a user name of 'administrator'. If your server is different
#' use this function to collect the new information. This updates
#' a file named ihub_connect.rdata to hold this information for the
#' other methods to use.
#' @return A file containing the connection information
#' @export

set_connection_info <- function(){
  server <- readline("What is the server name; http://")
  username <- readline("What is your username?")
  password <- readline("What is your password?")

print(paste("iHub Server:", server, " - iHub user name:", username, " - iHub password:", password))
connect <- list(ihub.server=server, ihub.username=username, ihub.password=password)
save(connect,file="ihub_connect.rdata")
}

#' Upload a file to Information Hub
#'
#' Upload a file from the R workspace to a location on
#' Information Hub volume. 
#' 
#' @examples
#' upload2ihub('readme.md','/Resources/','')
#' 
#' @return Response code from the Information Hub server
#' @param filename the name and file extension
#' @param folder the path on ihub volume to store file like /Resources/
#' @param overwrite true to overwrite existing file 
#' @export

upload2ihub <- function(filename,folder,overwrite){
require(curl)
require(jsonlite)
require(httr)

#SET LOGIN information
load("ihub_connect.rdata")
login <- list(
  username = connect$ihub.user,
  password = connect$ihub.password
)

#GET AUTHTOKEN - add verbose() as last encode for debugging
url.login <- paste("http://",connect$ihub.server,":8000/api/v2/login",sep="")
req <- httr::POST(url.login,
  body = login,
  encode = "form"
);
token <- paste(content(req)$authToken)

#UPLOAD FILE with authToken to a volume directory
#mime type is guessed from extension, otherwise add the mimetype
# upload_file("test.pdf", "application/pdf"))
# overwrite and mime not yet in code
url.upload <- paste("http://",connect$ihub.server,":8000/api/v2/files/upload",sep="")
req2 <- httr::POST(url.upload,
  httr::add_headers(
    "AuthToken" = token,
    "Content-Type" = "multipart/form-data",
	"Accept-Encoding" = "gzip, deflate",
    "Accept" = "application/json"
  ),
  body=list(name=paste(folder,filename,sep=""), file=upload_file(filename))
);
print(req2)
}


#' Get data set from Information Hub
#'
#' Download a data from a data object and
#' put it into a data frame. 
#' 
#' @examples
#' get_dataset('myDataFrame', '/Resources/Data Objects/')
#' 
#' @return A dataframe created from the values in the data set
#'   
#' @param dataframe the name of the new data frame
#' @param folder the path on ihub volume to store file like /Resources/
#' @export
get_dataset <- function(dataframe,folder){
require(curl)
require(jsonlite)
require(httr)

if (folder == '') { folder = '/Resources/' } 

load("ihub_connect.rdata")
#SET LOGIN INFORMATION
login <- list(
  username = connect$ihub.user,
  password = connect$ihub.password
)

#GET AUTH TOKEN - add verbose() as last encode for debugging
url.login <- paste("http://",connect$ihub.server,":8000/api/v2/login",sep="")
req <- httr::POST(url.login,	
  body = login,
  encode = "form"
);
token <- paste(content(req)$authToken)
myfolder = 
#GET DATA FILES with authToken - http://localhost:8000/api/v2/files?search=%2FResources%2F*.data
url.dataobjects <- paste("http://",connect$ihub.server,":8000/api/v2/files?search=",URLencode(folder),"*.data",sep="")
req2 <- httr::GET(url.dataobjects, add_headers(AuthToken = token))
json <- httr::content(req2, as = "text")
mydata1 <- fromJSON(json)
datafiles = NULL

for (i in 1:length(mydata1$itemList$file$name))
	datafiles[i] <- mydata1$itemList$file$name[i]
choicefile <- menu(datafiles, title = 'Select a data object file')

#USE mydata1$itemList$file$id[choicefile]

#GET DATA SETS with authToken - http://localhost:8000/api/v2/dataobjects/15000000100
url.datasets <- paste("http://",connect$ihub.server,":8000/api/v2/dataobjects/",mydata1$itemList$file$id[choicefile],sep="")
req3 <- httr::GET(url.datasets, add_headers(AuthToken = token))
json2 <- httr::content(req3, as = "text")
mydata2 <- fromJSON(json2)
datasets = NULL

for (i in 1:length(mydata2$GetMetaDataResponse$ArrayOfResultSetSchema$ResultSetSchema$ResultSetName))
	datasets[i] <- mydata2$GetMetaDataResponse$ArrayOfResultSetSchema$ResultSetSchema$ResultSetName[i]
choicedataset <- select.list(datasets, title = 'Select a data set to download')

#USE choicedataset

#DOWNLOAD DATA SETS with authToken - http://localhost:8000/api/v2/dataobjects/15000000100/Customers
url.data <- paste("http://",connect$ihub.server,":8000/api/v2/dataobjects/",mydata1$itemList$file$id[choicefile],"/",URLencode(choicedataset),sep="")
req4 <- httr::GET(url.data, add_headers(AuthToken = token))
json3 <- httr::content(req4, as = "text")
mydata3 <- fromJSON(json3)

assign(paste(dataframe), mydata3[['data']], envir=globalenv())
}